### DISCUSSION TITLE
How to bind stored procedure result to Python model?

### DISCUSSION THREAD

**Discussion by ManelBH** [sqlalchemy/sqlalchemy#9967] [Usage Questions]
I have a stored procedure that returns a table as a result. I want to execute a procedure and bind a Python model to the result. 
I tried something like:

`query = session.query(MyModel).from_statement(stmt)`

But SQLAlchemy compalins that the model has no tablename or no ID, problem is there isn't a table nor id column.
And in a more general sense I would expect SQLAlchemy being able to map any result table from an arbitrary statement to a matching model. 
Is there a way to do this?

**Comment by zzzeek** [comment]
the model has to be against the statement itself as a subquery like you see here:   https://docs.sqlalchemy.org/en/20/orm/nonstandard_mappings.html#mapping-a-class-against-arbitrary-subqueries

would need more details for further info

**Comment by ManelBH** [comment]
Thanks for the answer but in my case the procedure needs a parameter so I'm not sure how would this work. To add some more detail my procedure generates a report at a given date, that's why I need to pass a date parameter at runtime.

**Reply by zzzeek** [reply]
it would be awkward but if you use [bindparam()](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.bindparam) you could still do basic queries

```py
class X(Base):
    __table__ = select(func.myproc(bindparam('id'))).subquery()

objects = session.scalars(select(X), {"id": 5})
```

if this were a high value target it might be possible to use statement translation events to reroute WHERE clauses into the function, so that features like `session.get()` and `session.refresh()` could work.    seems likely not worth it.

having your stored procedure reworked as a view which could allow for a WHERE clause with a primary key column would be a lot more fluent.
